Prototype Managed QUIC support implementation

rzikm opened the pull request:
This PR adds initial version of almost C#-only QUIC implementation as an alternative to MsQuic based implementation. The work originated as my Master Thesis.

## Supported QUIC Protocol Features

This implementation is highly in WIP state. The features are at the draft-27 version. The goal was to obtain a minimal working
implementation able to reliably transmit data between client and server. This can be used to evaluate the viability of purely C# implementation for production releases. Most transport-unrelated features are
left unimplemented.

Currently implemented:

- Basic connection establishment
- Encryption
- TLS integration backed by [modified OpenSSL](https://github.com/openssl/openssl/pull/8797)
- Sending data, flow control, loss recovery, congestion control
- Stream/Connection termination
- Coalescing packets into a single UDP datagram

## Unsupported QUIC Protocol Features

A list of highlights of unimplemented protocol features follows:

- Connection migration
- Stateless reset
- 0-RTT data
- Server Preferred Address
- Version Negotiation
- Address Validation
- Encryption key updates

## OpenSSL integration

To function correctly, the implementation requires a custom branch of OpenSSL from Akamai
https://github.com/akamai/openssl/tree/OpenSSL_1_1_1g-quic. The implementation tries to detect that
a QUIC-supporting OpenSSL is present in the path. If not, then a mock implementation is used instead.

The mock implementation can be used for trying the implementation locally, but interoperation with other
QUIC implementations is, of course, possible only with the OpenSSL-based TLS. If you want to make sure your
implementation runs with OpenSSL, define the `DOTNETQUIC_OPENSSL` environment variable. This will
cause the implementation to throw if proper OpenSSL version cannot be loaded.

## Tracing and qvis

The implementation can produce traces that can be consumed by https://qvis.edm.uhasselt.be
visualizer. To collect traces, define `DOTNETQUIC_TRACE` environment variable to something nonempty. The traces will be
saved in the working directory in a file named _[timestamp]-[server|client].qvis_.

## Disabling encryption

Regardless of TLS used, you can circumvent encryption by defining the `DOTNETQUIC_NOENCRYPT` environment variable to something nonempty.

---

Ported from https://github.com/dotnet/runtime/pull/43948

rzikm committed:
Add sending ACK frames.

rzikm committed:
Split Frame related code from Connection to partial.

rzikm committed:
Use single PaddingFrame in test harness.

rzikm committed:
Use negotiated cipher suite.

rzikm committed:
Allow specifying length when writing varints.

rzikm committed:
First implementation of sending error frames.

rzikm committed:
Correctly send 1-RTT packets.

rzikm committed:
Remove some simple TODOs.

rzikm committed:
Add context class to store data during Send and Receive calls.

rzikm committed:
Add Remove functions to the RangeSet

rzikm committed:
Add outbound buffering implementation.

rzikm committed:
Add prototype InboundBuffer implementation.

rzikm committed:
Handle out of order and duplicate reception in InboundBuffer

rzikm committed:
Refactor retrieving peer transport parameters.

rzikm committed:
OutboundBuffer now respects MaxData.

rzikm committed:
Basic stream book-keeping - part I.

rzikm committed:
Send and handle HANDSHAKE_DONE frames.

rzikm committed:
Correctly initialize outbound streams with limits

rzikm committed:
Set supported ciphers explicitly.

rzikm committed:
Implement infrastructure for acking packets.

rzikm committed:
Refactored error handling and frame processing

rzikm committed:
First successful stream data transport.

rzikm committed:
Add tests for some error states for stream frames.

rzikm committed:
Migrated to signed types to avid casting everywhere.

rzikm committed:
Receiving and sending final size of the stream.

rzikm committed:
Use wildcard include to fix build problems.

rzikm committed:
Introduce a QuicSocketContext for background socket thread.

rzikm committed:
Allocate callbacks on unmanaged heap to prevent relocation.

rzikm committed:
Simplify connection creation for tests.

rzikm committed:
Fixes and partial public API implementation

rzikm committed:
Make use of OutboundBuffer threadsafe.

rzikm committed:
Make InboundBuffer threadsafe.

rzikm committed:
Trivial async read from ManagedQuicStream

rzikm committed:
More efficient background thread for sockets.

rzikm committed:
Renamed Epoch to PacketNumberSpace for beter clarity.

rzikm committed:
Fix small issues.

rzikm committed:
Hacky Recovery implementation from RFC.

rzikm committed:
Refactored recovery logic to separate file.

rzikm committed:
Refactor RecoveryController, add some tests.

rzikm committed:
Add tests for New Reno

rzikm committed:
Minor fixes

rzikm committed:
Use custom built OpenSSL from repository

rzikm committed:
Fixes from code review.

rzikm committed:
Fix failing unit test.

rzikm committed:
Limit sent packets by the congestion window.

rzikm committed:
Support for timeouts.

rzikm committed:
Recover from lost packets.

rzikm committed:
Fixed acking unreceived packets.

rzikm committed:
Fix exception when local endpoint is null

rzikm committed:
Fixed infinite cycles when trying to write stream frame.

rzikm committed:
Improvised implementation of all write methods on stream.

rzikm committed:
Refactored out of order stream data receive.

rzikm committed:
Blocking read methods and other stream improvements.

rzikm committed:
Fix some hangs when desposing of Quic* objects

rzikm committed:
Limiting amount of buffers used to buffer stream data.

rzikm committed:
Add simple object pooling.

rzikm committed:
Partially working graceful shutdown.

rzikm committed:
Added event tracing.

rzikm committed:
Improvements and bugfixes.

rzikm committed:
Fix Recovery controler.

rzikm committed:
Improve connection stability.

rzikm committed:
Implement probe timeout

rzikm committed:
Signal connected event for server connection.

rzikm committed:
Fix frame serialization in test harness.

rzikm committed:
Fix resending crypto frames.

rzikm committed:
Fix incorrect loss probe timeout.

rzikm committed:
Fix stream unit tests.

rzikm committed:
Improve logging packets in test harness.

rzikm committed:
Fix livelock during connection establishment.

rzikm committed:
Fix sending wrong ack delay.

rzikm committed:
Minor fixes

rzikm committed:
Fixed wrong RTT calculations.

rzikm committed:
Implement simple stream-level control flow.

rzikm committed:
Wired up simple connection-level flow control.

rzikm committed:
Refactoring, comments.

rzikm committed:
Small changes for profiling.

rzikm committed:
Implement basic ALPN.

rzikm committed:
Simple idle timeout.

rzikm committed:
Refactor congestion control algorithms into stateless strategy classes.

rzikm committed:
Drop packet protection keys when no longer needed.

Drop

rzikm committed:
Allow use of msquic via envirotnment variable.

rzikm committed:
Prepare passive key update.

rzikm committed:
Small changes in test harness to allow reading traces.

rzikm committed:
Fix incorrect value of active_connection_id_limit transport parameter
being sent.

rzikm committed:
Add more tracing events.

rzikm committed:
Add partial support for NEW_CONNECTION_ID frame.

rzikm committed:
Faster connection close.

rzikm committed:
Dispatch datagrams on server by remote endpoint.

rzikm committed:
Small refactor of tests.

rzikm committed:
Add stubs for unsupported frames.

rzikm committed:
Implement STOP_SENDING and STREAM_RESET frames.

rzikm committed:
Handling of tls alerts.

rzikm committed:
Fix some functional tests made for MsQuic.

rzikm committed:
Add more tests for STOP_SENDING and RESET_STREAM.

rzikm committed:
Implemented state transitions in send/recv stream parts.

rzikm committed:
Improve memory performance of InboundBuffers.

rzikm committed:
Make RangeSet.Range a struct.

rzikm committed:
Use ValueTask in QuicSocketContext.

rzikm committed:
Move to sync methods in socket contexts.

rzikm committed:
Attempt to reduce memory allocations.

rzikm committed:
Use msquic for msquic tests.

rzikm committed:
Discarding unread data for locally aborted streams.

rzikm committed:
Process only CONNECTION_CLOSE frames when closing.

rzikm committed:
Remove obsolete TODOs.

Remove obsolete TODOs

rzikm committed:
Discard incoming data if stream no longer wants to receive.

rzikm committed:
Close connection when server sends Token in Initial packet.

rzikm committed:
Split header protection and payload encryption.

rzikm committed:
Refactor away ugly direct reads in packet processing.

rzikm committed:
Remove obsolete TODOs.

Remove obsolete TODOs

rzikm committed:
Return SentPackets back to pool when dropping packet number space.

rzikm committed:
Threadsafe opening new streams.

rzikm committed:
Remove leak on stopped listener.

rzikm committed:
Remove unused using directives.

rzikm committed:
InboundBuffer.RequestAbort now aborts readers.

rzikm committed:
Drop all buffered data on abort write on OutboundBuffer.

rzikm committed:
Disposing of QuicStream.

rzikm committed:
Remove QuicErrorException.

rzikm committed:
Close connection when cannot retrieve transport parameters.

rzikm committed:
Fix interop with msquic.

rzikm committed:
Fix exception in socket context on Windows.

rzikm committed:
Remove allocations when creating header protection mask.

rzikm committed:
Small optimizations.

rzikm committed:
Reduced sending MaxStreamData frames.

rzikm committed:
Optimized sending and receiving, gaining more throughput.

rzikm committed:
Further optimizations.

rzikm committed:
Fixed encoding of packet numbers.

rzikm committed:
Optimized away some allocations when processing acks.

rzikm committed:
Small optimizations.

rzikm committed:
Log datagrams in NetEventSource only if particular levels are enabled.

rzikm committed:
Set local limits in constructor.

rzikm committed:
Use smaller size for packets to avoid fragmentation.

rzikm committed:
Fix closing connection.

rzikm committed:
Improved performance of the socket context code.

rzikm committed:
Fix deadlock when sending stream with only Fin.

rzikm committed:
Fix some of the compilation issues in runtime repo.

rzikm committed:
Fix throwing exception on conneciton errors

rzikm committed:
Partial support for ServerCertificate objects.

rzikm committed:
Remove allocating lambdas from OutboundBuffer.

rzikm committed:
Remove more allocations.

rzikm committed:
Improve stability in out-of-order scenarios.

rzikm committed:
Fixed platform dependent IPv6 constant for msquic.

rzikm committed:
Move to custom build native lib.

rzikm committed:
Abort blocked writers when stream is aborted.

rzikm committed:
Remove obsolete TODOs

rzikm committed:
Validate reserved bits.

rzikm committed:
Flush stream when endStream parameter is set.

rzikm committed:
Detect duplicate transport parameters.

Commit:
Build native lib as part of dotnet runtime.

rzikm committed:
Fixed System.Net.Quic.Native build for linux.

Commit:
Make PDBs in System.Net.Quic.Native optional in Win builds.

Commit:
Fixed runtime build issues.

Commit:
Fix unit tests compilations

rzikm committed:
Remove NetEventSource.Enter/Exit

rzikm committed:
Fix compilation.

rzikm committed:
Remove client-side per-UDP send allocations.

rzikm committed:
Transfer connections to single connection context after handshake completion.

rzikm committed:
WIP solving of exceptions caused by ICMP messages

rzikm committed:
Fix running unit tests.

rzikm committed:
Fail fast when no certificate supplied

rzikm committed:
Using SocketAsyncEventArgs to remove the exception

rzikm committed:
Propagating the exception from QuicSocketContext

rzikm committed:
Use SocketAsyncArgs at the server side too

rzikm committed:
Fix accessing disposed socket due to race condition

rzikm committed:
Disable transfer to single connection context, as that one does not work on windows

rzikm committed:
Minor cleanup of QuicSocketContext

rzikm committed:
Make stream creation threadsafe

rzikm committed:
Minor changes

rzikm committed:
Moved source code to revived System.Net.Quic solution

rzikm committed:
Add missing references

rzikm committed:
Add Copyright text

rzikm committed:
Fix runtime compilation

rzikm committed:
Fix ALPN negotiation

rzikm committed:
Add Flush() calls to unit tests

rzikm committed:
Improve stability (at the cost of more allocations)

The idea of connected sockets needs to be revisited and the exceptional
cases need to be more thought out

rzikm committed:
Add qlog tracing implementation

rzikm committed:
Improvements based on the collected traces

rzikm committed:
Some minor fixes and improvements

rzikm committed:
Add possibility to disable encryption

rzikm committed:
Reorganizing the solution a bit

rzikm committed:
Preparing for mock TLS implementation

rzikm committed:
Fix runtime build

rzikm committed:
Provide mock implementation of TLS

rzikm committed:
Fix asserts firing with unit tests

rzikm committed:
Remove custom openssl from the build

Commit:
Improve check for modified openssl

rzikm committed:
Fix sending probe packets when no ack eliciting packets are in flight

rzikm committed:
Renames and other minor improvements

rzikm committed:
Repeat sending HANDSHAKE_DONE until it is acknowledged

rzikm committed:
Fix clearing up recovery packet number space

rzikm committed:
More bug fixes

rzikm committed:
Minor improvements and fixes

rzikm committed:
Pacer implementation

rzikm committed:
Runtime build fixes

Fix compilation for OSX and FreeBSD

Fix System.Net.Quic.Unit.Tests compilation

Commit:
Don't use 0 port in remote address in tests

rzikm committed:
Fix not clearing out stackalloc'd memory when necessary

Commit:
Support for console trace

Commit:
Better QuicSocketContext

Commit:
Fix cleanup on server side

Commit:
Trivial check if we are application or flow-control limited

Commit:
Small fix for console trace

Commit:
Fixed freezing due to a race condition

Commit:
Renamed ConnectionContext into QuicConnectionContext

rzikm committed:
Fix compilation after rebase

rzikm committed:
Updated readme

rzikm committed:
Remove unused using directives

geoffkizer commented:
Thanks for posting @rzikm.... I will review and comment on any issues.

It looks like a couple commits from runtime/master got added in to your PR, likely because you rebased there and then pushed here, and these commits don't exist in runtimelabs yet. I will merge the latest from runtime/master in here and hopefully this will fix that.

rzikm commented:
Yes, I struggled a bit with creating a PR into a repo, since runtimelab and runtime are unrelated in Github's eyes.

Anyway, I am thinking if it would be better to squash those commits. from the multiple rebases and conflict resolutions, it is unlikely that an arbitrary commit in the middle is compilable, let alone functional. thoughts?

geoffkizer commented:
> Anyway, I am thinking if it would be better to squash all those 250 or so commits. from the multiple rebases and conflict resolutions, it is unlikely that an arbitrary commit in the middle is compilable, let alone functional. thoughts?

It's probably worth squashing at least to some degree. There's always a tradeoff here of maintaining the history for documentation purposes (e.g., see exactly what we changed to enable some feature) vs the cost of doing so and complexity of having so many commits.

If you want to squash it all to one commit, I'd be fine with that. If you want to squash it down to some set of steps (i.e. commits) that retains some useful history/context that we might case about preserving for future information, I'm fine with that too. Up to you.

geoffkizer commented:
BTW, quick note: You shouldn't need `#nullable enable` in most of these files (at least source files; test code is different) because it's enabled at the project level.

geoffkizer commented:
CI is failing with this error: .packages\microsoft.dotnet.build.tasks.targetframework.sdk\6.0.0-beta.20529.1\build\Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.targets(66,5): error MSB3202: (NETCORE_ENGINEERING_TELEMETRY=Build) The project file "..\..\System.Runtime.InteropServices.JavaScript/src/System.Runtime.InteropServices.JavaScript.csproj" was not found.

Do you know if that's related to this PR or not? It's possible that it's a pre-existing issue since this branch just got set up. I'll investigate once I get the merging stuff figured out, but thought I'd mention it now.

geoffkizer commented:
BTW, the merge will also bring over this change: https://github.com/dotnet/runtime/pull/44223

This should simplify handling of the "browser" build, but may cause merge conflicts with your PR, not sure...

geoffkizer commented on README.md:
```diff
@@ -1,8 +1,66 @@
+
+### Console
+
+To get trances printed to the console, define `DOTNETQUIC_TRACE` environment variable to `console`.
```
Nit: typo -- "trances"

geoffkizer commented on src/libraries/System.Net.Http/src/System.Net.Http.csproj:
```diff
@@ -5,6 +5,7 @@
     <DefineConstants>$(DefineConstants);HTTP_DLL</DefineConstants>
     <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX;$(NetCoreAppCurrent)-FreeBSD;$(NetCoreAppCurrent)-iOS;$(NetCoreAppCurrent)-tvOS;$(NetCoreAppCurrent)-Browser</TargetFrameworks>
     <Nullable>enable</Nullable>
+    <DefineConstants>FEATURE_QUIC_PUBLIC</DefineConstants>
```
What is this for?

rzikm replied:
a remnant of very early code where the classes were still internal. I will remove it.

geoffkizer commented:
It looks like there are a bunch of changes to clean up "using" statements in the MsQuic and Mock providers. We should remove these changes as it's likely to just lead to merge conflicts.

geoffkizer commented:
Similarly, there are some changes to System.Net.Http, and some changes to the shared QUIC code (like adding a QuicConnection constructor). These may be good and/or necessary changes, but we probably shouldn't be making them here. 

We should either make these changes in runtime/master or drop them. 

In general, we should minimize changes in this tree outside of the managed implementation proper. If they are good changes, they should go into runtime/master. And making those sort of changes here can lead to merge nightmares later.

rzikm committed:
Revert changes to made to System.Net.Http project

rzikm committed:
Revert changes to Mock and MsQuic implementations

geoffkizer commented:
Looks like there are a few issues with Quic functional tests. Do you know what's happening here?

Also, looks like there are a bunch of issues with Http functional tests -- I'm guessing there is some basic issue here, haven't investigated though...

rzikm commented:
> Also, looks like there are a bunch of issues with Http functional tests -- I'm guessing there is some basic issue here, haven't investigated though...
That is baffling, I have reverted all changes in the System.Net.Http subtree, so it should behave exactly as in upstream

> Looks like there are a few issues with Quic functional tests. Do you know what's happening here?
I will try to investigate sometime soon

geoffkizer commented:
The reason the Http functional tests are failing seems to be that you've changed `QuicImplementationProviders.Default` to be the managed provider, and it returns true for `IsSupported`. There are some HTTP version selection tests that run conditionally based on this, and try to actually verify the expected version selection behavior by issuing HTTP requests against the loopback server.

It's not entirely clear to me whether these tests are actually correct or not -- some of them look a little wonky. But regardless, they weren't running in CI before because `QuicImplementationProviders.Default.IsSupported` was returning false, and now it's returning true, so they are running.

geoffkizer commented:
It's probably fine to change the default QUIC provider to the managed provider in this branch -- after all that's the point of this branch.

But what seems undesirable is that, when OpenSsl is not available, the managed provider will fall back to its mock TLS implementation, which means that code (like these tests) will think they can actually use QUIC when what they are actually getting is mock TLS QUIC.

I think we should separate out the mock TLS handling into a separate QUIC provider. That is, we should have:
(1) A managed QUIC provider that does real QUIC. This can be the default. This depends on OpenSsl, so it will return false from IsSupported if it can't find OpenSsl.
(2) A mock TLS managed QUIC provider -- something like `ManagedMockTls`. This doesn't depend on OpenSsl, so should always return true for IsSupported.

And we should run tests against both of these. The real managed provider won't run in CI, at least for now, because there's no OpenSsl. But at least the Mock TLS provider will run, which is valuable.

geoffkizer commented:
BTW, I verified that changing the default provider back to MsQuic makes the Http tests pass, because Default.IsSupported returns false now.

geoffkizer commented on src/libraries/System.Net.Http/src/Resources/SR.Designer.cs:
```diff
@@ -0,0 +1,127 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
```
Not sure how this file got added, but it shouldn't be. It's autogenerated during the build.

rzikm replied:
removed

geoffkizer commented on src/libraries/System.Net.Http/src/Resources/Strings.Designer.cs:
```diff
@@ -0,0 +1,925 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
```
Not sure how this file got added, but it shouldn't be. It's autogenerated during the build.

rzikm replied:
removed

geoffkizer commented on src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:
```diff
@@ -46,12 +46,18 @@ internal QuicListener CreateQuicListener()
         }
 
-        internal async Task RunClientServer(Func<QuicConnection, Task> clientFunction, Func<QuicConnection, Task> serverFunction, int millisecondsTimeout = 10_000)
+        internal async Task RunClientServer(Func<QuicConnection, Task> clientFunction, Func<QuicConnection, Task> serverFunction, int millisecondsTimeout = 10_000_000)
```
This is very long -- is it necessary? Increasing it a bit seems fine, but increasing it so much seems bad.

rzikm replied:
this was a mistake, I temporarily raised the timeout when debugging, I will remove it

geoffkizer commented:
I've manually ported the QUIC API changes back to dotnet/runtime. See here: https://github.com/dotnet/runtime/pull/44380

Once this gets in, I will merge it to the feature branch here and we should be able to rebase and get rid of the API changes in this PR.

geoffkizer commented:
FWIW, I can't easily reproduce any of the QUIC test failures locally. They seem to happen pretty consistently in CI. This usually means that the failures only happen under high CPU load, as occurs in CI.

geoffkizer commented:
Though I do notice, the functional tests take much longer than they used to with just the mock provider -- ~60s on my machine vs something like 5-10s with just the mock provider. 

Any idea why the tests take so long?

geoffkizer commented:
We should enable the new stream conformance tests for the managed provider. They are here: https://github.com/dotnet/runtimelab/blob/feature/ManagedQuic/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs

I tried doing this locally, but unfortunately I got a bunch of errors and some test hangs. I'm guessing the flush behavior on managed QuicStream is causing problems.

We may want to just add auto-flushing on every stream write, as a way to get these tests passing.

geoffkizer commented:
Also re conformance tests: Several tests are currently disabled because of the non-standard behavior of QuicStream on dispose. That's fine for now, but the plan is to fix the behavior of QuicStream on dispose to align with standard stream behavior. We should make this change for managed QuicStream and then re-enable these  tests.

I recently merged a PR to do this for the mock provider, see here: https://github.com/dotnet/runtime/pull/44282

geoffkizer commented:
We should also enable some of the HTTP3 tests in System.Net.Http to run against the managed provider.

The tests are here: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L3008

However, we should address issues with the stream conformance tests before attempting this.

geoffkizer commented:
Here's the change to enable stream conformance tests: https://github.com/geoffkizer/runtimelab/tree/addstreamconformancetests

stephentoub commented:
> I'm guessing the flush behavior on managed QuicStream is causing problems.

There's a property you can override on the test base class to say that writes require flushing.

geoffkizer commented:
> There's a property you can override on the test base class to say that writes require flushing.

Good point, thanks. Unfortunately, this doesn't seem to help. There's some other issue going on here.

rzikm committed:
Another Managed quic provider with Mock tls

rzikm committed:
Remove autogenerated files from the repo

rzikm committed:
Remove unwanted commited change

rzikm committed:
Add stream conformance tests for Managed implementation

rzikm commented:
I added `ManagedMockTls` provider, which always uses the mocked tls, the `Managed` one always uses OpenSSL (and so `IsSupported` should return false now)

> Though I do notice, the functional tests take much longer than they used to with just the mock provider -- ~60s on my machine vs ~3s with just the mock provider.
> 
> Any idea why the tests take so long?

I noticed that QuicStreamTests.LargeDataSentAndReceived takes awfully long time, since it sends 16MB of data and repeats that 100 times. Although it takes way longer than it should in my estimation. I am a bit behind on my thesis so I don't have time to investigate it further yet.

I also added the Stream conformance tests, but for some reason it cannot create the connected stream pairs, there are some OperationCancelledExceptions thrown around but I was unable to find the root cause. I will try to find time later next week to investigate.

rzikm committed:
Small stream conformance behavior improvement

geoffkizer commented on src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs:
```diff
@@ -45,10 +64,14 @@ protected override async Task<StreamPair> CreateConnectedStreamsAsync()
-                new SslServerAuthenticationOptions { ApplicationProtocols = new List<SslApplicationProtocol> { protocol } });
+            QuicListener listener = new QuicListener(provider, new QuicListenerOptions()
+            {
+                ListenEndPoint=new IPEndPoint(IPAddress.Loopback, 0),
```
Nit: spacing

geoffkizer commented on src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamConnectedStreamConformanceTests.cs:
```diff
@@ -22,6 +22,25 @@ public sealed class MsQuicQuicStreamConformanceTests : QuicStreamConformanceTest
+    public abstract class ManagedQuicStreamConformanceTestsBase : QuicStreamConformanceTests
+    {
+        protected override bool FlushRequiredToWriteData => true;
+        protected override bool FlushGuaranteesAllDataWritten => false;
```
Is this actually the case? What does Flush guarantee then, if it doesn't guarantee that all data is written?

geoffkizer replied:
To clarify the use of this flag: It's intended for something like a compression stream, where it may not be possible to actually write out all the provided input data because you've got a partial byte that can't be immediately flushed.

I don't think we should be using it for QuicStream.

stephentoub replied:
> What does Flush guarantee then, if it doesn't guarantee that all data is written?

For reference, the only two places we currently set this to false are when testing the compression streams (e.g. DeflateStream) and TranscodingStream, where there may be partial data left in the internal buffer (e.g. half a UTF8 character) that can't be written out until more data arrives to complete what's in the buffer.  I'm also surprised to see it set to false here.

rzikm replied:
Maybe I misunderstood that one, or even what Flush() should do. Currently, flush makes the data available to the background processing thread which will send them when possible, so technically when flush returns the other side did not necessarily receive the data yet.

stephentoub replied:
With zero additional interaction with the Stream, does the Flush call guarantee that all data previously written to the Stream will eventually reach and be readable by the other end of the connection?

geoffkizer replied:
> Currently, flush makes the data available to the background processing thread which will send them when possible, so technically when flush returns the other side did not necessarily receive the data yet.

Yes, that's the correct behavior for Flush, and you should be able to set FlushGuaranteesAllDataWritten to true.

geoffkizer commented on README.md:
```diff
@@ -1,8 +1,65 @@
+
+To function correctly, the implementation requires a custom branch of OpenSSL from Akamai
+https://github.com/akamai/openssl/tree/OpenSSL_1_1_1g-quic. The implementation tries to detect that
+a QUIC-supporting OpenSSL is present in the path. If not, then a mock implementation is used
```
Can we update this to reflect the two providers?

geoffkizer commented:
I've posted a merge PR: https://github.com/dotnet/runtimelab/pull/318

This should include the minor API changes from the PR: https://github.com/dotnet/runtime/pull/44380, so those can now be removed from this PR.

Note this will also pick up this PR: https://github.com/dotnet/runtime/pull/44282, which moves the disabled stream conformance tests to be MsQuic-specific. We can also disable them for the managed provider if necessary, though it would be nice to re-enable them before long. (The issue here is that Dispose should cause a graceful shutdown of the send stream, which it's not currently in MsQuic, and wasn't in the mock provider until I fixed it in that PR.)

geoffkizer commented:
> I noticed that QuicStreamTests.LargeDataSentAndReceived takes awfully long time, since it sends 16MB of data and repeats that 100 times. Although it takes way longer than it should in my estimation. I am a bit behind on my thesis so I don't have time to investigate it further yet.

I don't think we need to fix this before merging the PR.

geoffkizer commented:
I looked at the stream conformance stuff, and here's what seems to be happening:

The logic in `QuicStreamConformanceTests.CreateConnectedStreamsAsync` is creating the streams using OpenBidirectionalStream(). The problem here is that this doesn't necessarily send anything on the wire, until you do at least one Write and Flush. So the server-side code here is hanging on AcceptStreamAsync because no stream is actually created on the wire.

(Unfortunately, the mock provider does create the stream immediately, without any Write or Flush happening. This seems problematic.)

I was hoping that calling Flush on the newly-created stream would cause it to create the stream on the wire with a 0-length STREAM frame, but that didn't work. That said, I think it *should* work, and if it did it would solve the problems here.

This touches on the larger issue of how flushing works in general; we need to sort that out eventually, but not right now.

nibanks commented:
> This touches on the larger issue of how flushing works in general; we need to sort that out eventually, but not right now.

MsQuic handles this in our [StreamStart](https://github.com/microsoft/msquic/blob/main/src/inc/msquic.h#L1013) function by taking [QUIC_STREAM_START_FLAGS](https://github.com/microsoft/msquic/blob/main/src/inc/msquic.h#L144), one of which is [QUIC_STREAM_START_FLAG_IMMEDIATE](https://github.com/microsoft/msquic/blob/main/src/inc/msquic.h#L147). This flag changes the default behavior (don't send anything until the app does a send) to immediately indicate the stream (via an empty STREAM frame) to the peer.

geoffkizer commented:
I merged recent changes from master, including the QUIC API updates: https://github.com/dotnet/runtimelab/pull/318

Commit:
hack stream conformance to force stream creation

geoffkizer commented:
I didn't actually mean to push that commit here, but I suppose there's no real harm in it :)

This updates the QUIC stream conformance stuff to force creation of the QUIC stream on the wire by just writing/reading a byte.

When I did this, I saw a couple asserts about stream.CanRead/CanWrite; I'm guessing these are related to your recent change here. So I commented these out temporarily. Now I'm hitting the assert in SendStream.Enqueue: `Debug.Assert(!SizeKnown);`

rzikm commented:
> Now I'm hitting the assert in SendStream.Enqueue: Debug.Assert(!SizeKnown);

That implies an attempt to write after the end of the stream has been marked.

rzikm committed:
Fix after rebase

rzikm committed:
Clean up commented asserts

rzikm committed:
Updated readme

rzikm commented:
Rebased,

Also, I tried a quick look on the asserts in tests, and it seems that only some tests are causing assertions
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ReadWrite_Success
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ReadWrite_Success_Large

Additionally, these seem to hang
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ConcurrentBidirectionalReadsWrites_Success
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ReadAsync_DuringReadAsync_ThrowsIfUnsupported
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.CopyToAsync_AllDataCopied_Large.CopyToAsync_AllDataCopied_Large

A lot other tests are failing, because the behavior likely differs from what is expected, I hope to get to it this week.

rzikm committed:
Fix synchronous Read not blocking until data available

rzikm committed:
Throw if the cancellation token is pre-cancelled

rzikm commented:
After fixing some issues (Not blocking on synchronous Read), I see no more asserts in tests 👍 , although I am a bit suspicious about it. Anyway, I fixed also some other tests, and now all tests pass, except for the following, which hang/take too long.

System.Net.Quic.Tests.QuicStreamTests_ManagedMockTlsProvider.LargeDataSentAndReceived
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.CopyToAsync_AllDataCopied_Large
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ReadWrite_Success_Large
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ZeroByteRead_BlocksUntilDataAvailableOrNops
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ConcurrentBidirectionalReadsWrites_Success
System.Net.Quic.Tests.ManagedMockTlsQuicQuicStreamConformanceTests.ReadAsync_DuringReadAsync_ThrowsIfUnsupported

Of course, same for Managed provider.

The investigation will continue another day.

geoffkizer commented:
That's awesome, thanks.

If some tests are just taking too long (i.e. perf issue, not functional issue) then we can disable the test for now. We know we have perf work to do here.

rzikm committed:
Fix failing unit tests

rzikm committed:
Fix possible access of disposed socket

geoffkizer commented:
@rzikm I'm seeing a bunch of test failures now in CI -- do you know what's going on here?

rzikm commented:
> @rzikm I'm seeing a bunch of test failures now in CI -- do you know what's going on here?

There is one assert, that one would cause sending of an empty frame (I want to filter out any sending before the execution reaches that point because of how qlog tracing is implemented). But it should not break anything significant in Release.

As for the other failures, my first guess is that there is a race condition possibly causing deadlock. I was fixing some ObjectDisposedExceptions this morning and did not see these locally.

rzikm committed:
Code organization, removed useless files

DemiMarie commented:
Instead of relying on a patched OpenSSL, another option would be to embed a copy of BoringSSL or rustls, both of which already support QUIC.

rzikm committed:
Minor changes

rzikm committed:
Use function pointers for OpenSSL integration

rzikm committed:
Remove explicit payload length argument from CryptoSeal methods

rzikm committed:
Use expected packet number instead of last acked one in crypto seals

rzikm committed:
Remove async method state machine for accepting connections

rzikm committed:
Cosmetic changes

rzikm committed:
Fix data race leading to deadlock

rzikm committed:
Dispose async now waits for all resources to be cleaned

rzikm committed:
Hack workaround for object disposed exception

rzikm committed:
Fix leaking buffers in SendStream

rzikm committed:
Pacer improvements

rzikm committed:
Fix overflow in pacer allowance calculation

rzikm committed:
Reduce padding in packets

rzikm committed:
Fix failing unit tests

rzikm committed:
Improved application limitting detection

rzikm committed:
Catch ChannelClosedException

rzikm committed:
Use ODCID in the qlog trace name

rzikm committed:
Fix pacing timer firing during handshake

rzikm committed:
Try fixing failing CI runs by disabling parallelism for running Managed QUIC tests

rzikm commented:
/azp run runtime-libraries-coreclr outerloop

rzikm committed:
Avoid OperationCancelledExceptions in QuicConnectionContext

rzikm committed:
Minor improvements

rzikm commented:
@geoffkizer, In the last test run, all QUIC tests passed, the only one failed test is in Sockets and HTTP2

https://helix.dot.net/api/2019-06-17/jobs/9e1d5026-ee81-4786-99cb-bfcfd335697e/workitems/System.Net.Http.Functional.Tests/console

This feels like random unrelated fail. Could you rerun the pipeline?

rzikm committed:
Reduce memory footprint

rzikm committed:
Fix checking incorrect limit for protocol violation

rzikm committed:
Fix failing unit test

rzikm committed:
Avoid growing congestion window when limited by flow control or application

rzikm committed:
WIP CUBIC implementation

rzikm committed:
Minor comments

geoffkizer commented:
Sorry, lost track of this. It looks like there are only minor failures in CI. Do you know what the issue here is?

rzikm commented:
I did not have much time to spend on it because I was too busy writing the master thesis text. But there seems to be some race condition leading to deadlock/livelock of either client or server. Sadly, it does not seem to be isolated to a single test that could be disabled.

What I find quite strange is that the probability of failure seems to depend on the order in which the tests are run. I will try to investigate again this week when I have time.

DemiMarie commented:
> I did not have much time to spend on it because I was too busy writing the master thesis text. But there seems to be some race condition leading to deadlock/livelock of either client or server. Sadly, it does not seem to be isolated to a single test that could be disabled.
> 
> What I find quite strange is that the probability of failure seems to depend on the order in which the tests are run. I will try to investigate again this week when I have time.

The Linux `netem` framework (part of `tc`) is fantastic for this.  I used it to find https://github.com/quinn-rs/quinn/pull/692, which was a bug in another QUIC implementation with similar symptoms.

rzikm commented:
> The Linux `netem` framework (part of `tc`) is fantastic for this. I used it to find [quinn-rs/quinn#692](https://github.com/quinn-rs/quinn/pull/692), which was a bug in another QUIC implementation with similar symptoms.

I am working primarily on Windows now, but simulating arbitrary network issues might help tracking this down, or at least reproduce it on my machine (as far as I know the tests fail only on CI). Thanks for the idea.

DemiMarie commented:
> > The Linux `netem` framework (part of `tc`) is fantastic for this. I used it to find [quinn-rs/quinn#692](https://github.com/quinn-rs/quinn/pull/692), which was a bug in another QUIC implementation with similar symptoms.
> 
> I am working primarily on Windows now, but simulating arbitrary network issues might help tracking this down, or at least reproduce it on my machine (as far as I know the tests fail only on CI). Thanks for the idea.

You’re welcome!

rzikm committed:
Use SafeHandles for SSL structure

rzikm committed:
Add unit test for resending FIN bit

ladeak commented:
@rzikm is there a reason for this managed QUIC PR not being merged even to the feature branch? I would be interested in learning about the performance of a managed implementation vs. msquic and other implementations?

rzikm commented:
> @rzikm is there a reason for this managed QUIC PR not being merged even to the feature branch? I would be interested in learning about the performance of a managed implementation vs. msquic and other implementations?

Hi, Thanks for the interest. The managed QUIC implementation here was my [master thesis](https://dspace.cuni.cz/bitstream/handle/20.500.11956/124659/120381248.pdf?sequence=1&isAllowed=y) and the goal was to investigate perf characteristics of managed implementation vs. relying on a native library (MsQuic in this case). There were some expectations of improvements due to not having to cross managed/native barrier and adapting to the native library's threading model. 

The thesis contains some measurement results, but looking back at it, it does not measure the same usage pattern as in e.g. HTTP/3 (benchmark uses long-lived streams instead of one-stream-per-request), so the numbers in the thesis are of limited value. The managed implementation *might* beat MsQuic-based implementation on some specific scenarios, but once you put enough load on it, it will degrade rapidly.

After all, as one university student, it is hard to outclass a dedicated team of networking engineers working on the library full-time :)

As for why it is not merged... it is hard to make it work without significant external support. It might be possible to make it work on Windows (Schannel API is flexible enough to implement QUIC on top of it), but on Linux and others, MsQuic brings their own TLS library (fork of OpenSSL with added QUIC APIs). At the time I wrote the thesiss, it relied on the QUIC-enabled libssl fork being available on the system, but most of the work was done with mocking out the TLS layer entirely.
